src/solve/Solver.hpp

```diff
@@ -522,176 +527,200 @@ class Solver
 #ifdef __cpp_lib_execution
                 std::execution::par_unseq,
 #endif
                 R_sorted.begin(), R_sorted.end(),
-                [](auto &a, auto &b)
-                { return a.pair_hash < b.pair_hash; });
-            /*std::sort(R_sorted.begin(), R_sorted.end(),
-                      [](auto &a, auto &b){ return a.pair_hash < b.pair_hash; });*/
-
+                [](const T1_Match &a, const T1_Match &b)
+                {
+                    return a.pair_hash < b.pair_hash;
+                });
             timings_.t2_sort_short_list += sub_timer.stop();
 
-            for (size_t j = 0; j < R_sorted.size(); ++j)
+            // --- build hash_to_index: reduced hash -> first index in R_sorted with that reduced value ---
+            sub_timer.start();
+
+            std::fill(hash_to_index.begin(), hash_to_index.end(), INVALID_INDEX);
+            assert(R_sorted.size() < INVALID_INDEX); // ensure we don't overflow uint16_t
+            for (uint16_t j = 0; j < numeric_cast<uint16_t>(R_sorted.size()); ++j)
             {
-                uint32_t hash_reduced = R_sorted[j].pair_hash >> (num_k_bits - HASHES_BITMASK_SIZE_BITS);
-                int slot = hash_reduced >> 5;
-                int bit = hash_reduced & 31;
-                hashes_bitmask[slot] |= (1 << bit);
+                uint32_t reduced = R_sorted[j].pair_hash >> REDUCE_SHIFT;
+                if (hash_to_index[reduced] == INVALID_INDEX)
+                {
+                    // Because R_sorted is sorted, this will be the lowest index for this reduced value.
+                    hash_to_index[reduced] = j;
+                }
+                #ifdef DEBUG_VERIFY
+                else {
+                    num_R_collisions++;
+                }
+                #endif
             }
+            #ifdef DEBUG_VERIFY
+            std::cout << "Num R collisions: " << num_R_collisions << std::endl;
+            #endif
+
+            timings_.t2_sort_short_list += sub_timer.stop(); // reuse this bucket for index build time
 
             const auto &L_list = t1_match_groups[t1_group_l];
-            ProofCore proof_core(params_);
             uint32_t num_match_keys = 1u << num_T2_match_key_bits;
 
-            // --- use C++20 parallel for to speed up the processing of L_list ---
-            sub_timer.start();
+            auto &out = t2_matches[t2_group];
+            std::mutex out_mutex;
 
-            // prepare shared list and mutex
-            L_short_list.clear();
-            L_short_list.reserve(L_list.size() * 2);
-            std::mutex list_mutex;
+            // --- parallel processing of unsorted L_list ---
+            sub_timer.start();
 
-            // parallel loop over each L element
-            parallel_for_range(L_list.begin(), L_list.end(), [&](const T1_Match &lm)
+            parallel_for_range(L_list.begin(), L_list.end(),
+                               [&](const T1_Match &lm)
                                {
```

Comment:
the block below here is extremely difficult to review, since the indentation and formatting changed together with logic changes.

Do you think it would be reasonable to restore the formatting?